sign/dilithium/mode3/internal/dilithium.go

```diff
@@ -0,0 +1,410 @@
+	h.Write(packedPk[:]) // nolint:errcheck
+	h.Read(sk.tr[:])     // nolint:errcheck
+
+	// Finish cache of public key
+	pk.tr = &sk.tr
+
+	return &pk, &sk
+}
+
+// Computes t0 and t1 from sk.s1h, sk.s2 and sk.A.
+func (sk *PrivateKey) computeT0andT1(t0, t1 *VecK) {
+	var t VecK
+
+	// Set t to A s₁ + s₂
+	for i := 0; i < K; i++ {
+		PolyDotHat(&t[i], &sk.A[i], &sk.s1h)
+		t[i].ReduceLe2Q()
+		t[i].InvNTT()
+	}
+	t.Add(&t, &sk.s2)
+	t.Normalize()
+
+	// Compute t₀, t₁ = Power2Round(t)
+	t.Power2Round(t0, t1)
+}
+
+// NewKeyFromSeed derives a public/private key pair using the given seed.
+func NewKeyFromSeed(seed *[common.SeedSize]byte) (*PublicKey, *PrivateKey) {
+	var buf [96]byte
+	h := shake.NewShake128()
+	h.Write(seed[:]) // nolint:errcheck
+	h.Read(buf[:])   // nolint:errcheck
+	return NewKeyFromExpandedSeed(&buf)
+}
+
+// Verify checks whether the given signature by pk on msg is valid.
+func Verify(pk *PublicKey, msg []byte, signature []byte) bool {
+	var sig unpackedSignature
+	var mu [48]byte
+	var zh VecL
+	var Az, Az2dct1, w1 VecK
+	var ch, cp common.Poly
+
+	// Note that Unpack() checked whether ‖z‖_∞ < γ₁ - β
+	// and ensured that there at most ω ones in pk.hint.
+	if !sig.Unpack(signature) {
+		return false
+	}
+
+	// μ = CRH(tr ‖ msg)
+	h := shake.NewShake256()
+	h.Write(pk.tr[:]) // nolint:errcheck
+	h.Write(msg)      // nolint:errcheck
+	h.Read(mu[:])     // nolint:errcheck
+
+	// Compute Az
+	zh = sig.z
+	zh.NTT()
+
+	for i := 0; i < K; i++ {
+		PolyDotHat(&Az[i], &pk.A[i], &zh)
+	}
+
+	// Next, we compute Az - 2ᵈ·c·t₁.
+	// Note that the coefficients of t₁ are bounded by 256 = 2⁹,
+	// so the coefficients of Az2dct1 will bounded by 2⁹⁺ᵈ = 2²³ < 2q,
+	// which is small enough for NTT().
+	Az2dct1.MulBy2toD(&pk.t1)
+	Az2dct1.NTT()
+	ch = sig.c
+	ch.NTT()
+	for i := 0; i < K; i++ {
+		Az2dct1[i].MulHat(&Az2dct1[i], &ch)
+	}
+	Az2dct1.Sub(&Az, &Az2dct1)
+	Az2dct1.ReduceLe2Q()
+	Az2dct1.InvNTT()
+	Az2dct1.NormalizeAssumingLe2Q()
+
+	// UseHint(pk.hint, Az - 2ᵈ·c·t₁)
+	//    = UseHint(pk.hint, w - c·s₂ + c·t₀)
+	//    = UseHint(pk.hint, r + c·t₀)
+	//    = r₁ = w₁.
+	w1.UseHint(&Az2dct1, &sig.hint)
+
+	// c' = H(μ, w₁)
+	PolyDeriveUniformB60(&cp, &mu, &w1)
+	return sig.c == cp
+}
+
+// SignTo signs the given message and writes the signature into signature.
+func SignTo(sk *PrivateKey, msg []byte, signature []byte) {
+	var mu, rhop [48]byte
+	var y, yh VecL
+	var w, w0, w1, w0mcs2, ct0, w0mcs2pct0 VecK
+	var ch common.Poly
+	var yNonce uint16
+	var sig unpackedSignature
+
+	if len(signature) < SignatureSize {
+		panic("Signature does not fit in that byteslice")
+	}
+
+	//  μ = CRH(tr ‖ msg)
+	h := shake.NewShake256()
+	h.Write(sk.tr[:]) // nolint:errcheck
+	h.Write(msg)      // nolint:errcheck
+	h.Read(mu[:])     // nolint:errcheck
+
+	// ρ' = CRH(μ ‖ key)
+	h.Reset()
+	h.Write(sk.key[:]) // nolint:errcheck
+	h.Write(mu[:])     // nolint:errcheck
+	h.Read(rhop[:])    // nolint:errcheck
+
+	// Main rejection loop
+	for {
```

Comment:
This for loop reminds me a bug on ecc arithmetic that caused CPU consumption. In order to prevent that kind of bug, what if the signature returns error after trying N times. Of course, it is expected that signature succeeds much earlier than N times.
We can fix this in another PR.

Reply:
To get the probability of a false failure below `2^-128`, we need `log(2^-128)/ log(1/7) ~ 151` iterations.

Reply:
I've added this to the current branch as it's a minor change.